### PR TITLE
Add missing docs

### LINKED ISSUE #569: interledger-service API review for v1.0-rc
interledger-service public items marked with `x` are documented. Non-exported items are not taken into account. Follows format from #561. 

Part of #557.

- [x] `username` (struct)
- [x] `trace`(module): Does this need to be exported? It just contains implementations to the service traits for `Instrumented`
- [x] `Account` (trait)
   - Should we document each function's expected output? Their names are pretty self-explanatory.
- [x] `IncomingRequest` (struct)
   - Maybe we should add comments to each field of the struct as well?
- [x] `OutgoingRequest` (struct)
   - Maybe we should add comments to each field of the struct as well?
- [x] `IncomingService` (trait):
   - Should we document the expected behavior of `handle_request`?
- [x] `OutgoingService` (trait):
   - Should we document the expected behavior of `send_request`?
- [x] BoxedIlpFuture (type)
- [x] `AccountStore` (trait):
   - Should we document each function's expected output? Their names are pretty self-explanatory.
- [x] `incoming_service_fn` / `outgoing_service_fn`
- [x] `ServiceFn`
- [x] `WrappedService`: 
   - (nit) we can rename the generic parameters for the `wrap_incoming` and `wrap_outgoing` functions to be `I` and `O` for consistency, instead of `IO`
- [x] `AddressStore`: 
   - Should be filled to explain why this functionality is needed/used, short explanation: when a parent account is added the node's address must be updated to reflect the address assigned to it by the parent

### LINKED ISSUE #570: interledger-btp API review for v1.0-rc
`interledger-btp` public items marked with x are documented. Non-exported items are not taken into account. Follows format from #561.

Part of #557.

- [x] `BtpAccount` (trait)
  - Should the trait functions also be documented? Seem a bit self explanatory from the name
- [x] `BtpStore` (trait)
- [x] `connect_client` (fn)
- [x] `connect_to_service_account` (fn)
- [x] `parse_btp_url` (fn)
   - Should this be removed? We do not use it anywhere, and just manually remove the `btp+` prefix here: https://github.com/interledger-rs/interledger-rs/blob/master/crates/interledger-btp/src/client.rs#L63
- [x] `btp_service_as_filter` (fn)
  - This is documented, but the comment is outdated. Needs to be updated to show that it converts the provided btp service to a warp filter
- [x] `BtpOutgoingService` (struct)
- [x] `BtpService` (struct)

### LINKED ISSUE #571: interledger-ccp API review for v1.0-rc
`interledger-ccp` public items marked with x are documented. Non-exported items are not taken into account. Follows format from #561.

Part of #557.

- [x] `CcpRoutingAccount` (trait)
- [x] `RouteManagerStore` (trait):
   - Should this be renamed to `CcpRoutingStore`? That would make it consistent with how the other Account/Store pairs are named.
- [x] `RoutingRelation`
- [x] `Mode`
   - We never use `Idle` mode. Should we remove Mode altogether and assume `Sync`? https://github.com/interledger/rfcs/pull/455 is closed and seems stale atm, so I am not sure what's the path towards CCP standardization
- [x] `RouteControlRequest`
   - Should `to_prepare` be implemented as `Prepare::from(src: RouteControlRequest)`?
- [x] `RouteProp`
   - Should we convert it to `pub(crate)`?
- [x] `CcpRouteManager` 
- [x] `CcpRouteManagerBuilder`: Does this need to be documented? Its fields are docuemnted, and given that it's a builder its functionality should be intuitive.

### LINKED ISSUE #572: interledger-router API review for v1.0-rc
`interledger-router` public items marked with x are documented. Non-exported items are not taken into account. Follows format from #561.

Part of #557.

- [x] `Router` (struct)
- [x] `RouterStore` (trait)

### LINKED ISSUE #573: interledger-ildcp API review for v1.0-rc
`interledger-ildcp` public items marked with x are documented. Non-exported items are not taken into account. Follows format from #561.

Part of #557.

- [x] `get_ildcp_info` (fn)
- [x] `IldcpService` (struct)
   - `handle_request` should be documented
- [x] `packet::IldcpRequest`
   - Can we move the `to_prepare` implementation inside the `From` implementation?
- [x] `packet::IldcpResponse`

### LINKED ISSUE #574: interledger-http API review for v1.0-rc
interledger-http public items marked with `x` are documented. Non-exported items are not taken into account. Follows format from #561. 

Part of #557.

- [x] `HttpAccount` (trait)
- [x] `HttpStore` (trait)
- [x] `deserialize_json` (fn)
- [x] `HttpServer` (struct)
   - `as_filter()`: Should we put some docs on the exact API endpoints this exposes or point to the spec?
- [x] `HttpClientService` (struct)
   - Maybe document that `send_request` serializes (& oer encodes) the ILP packet before sending it to the next service?
- [x] `error` (module)
   - Should we extract this to a separate crate? Seems generic enough, we discussed this another time.

### LINKED ISSUE #575: interledger-stream API review for v1.0-rc
interledger-stream public items marked with `x` are documented. Non-exported items are not taken into account. Follows format from #561. 

Part of #557.

- [x] StreamDelivery (struct)
- [x] `send_money` (fn)
- [x] `SendMoneyFuture` (struct)
  - Not public, but its implementation and fields must be documented in detail since it's our main low-level payments protocol and may be a prime target for future optimizations etc.
- [x] `Error` (enum)
  - Do we need this? It's only imported in the SPSP Error wrapper enum, but that variant does not seem to get hit ever
- [x] `ConnectionGenerator` (struct)
- [x] `PaymentNotification (struct)
- [x] `StreamNotificationStore` (trait)
- [x] `StreamReceiverService` (struct)

### LINKED ISSUE #576: interledger-settlement API review for v1.0-rc
interledger-settlement public items marked with `x` are documented. Non-exported items are not taken into account. Follows format from #561. 

Part of #557.

- [x] `api` (module)
- [x] `client::SettlementClient` struct (need to also document its `send_settlement` method)
- [x] `SettlementMessageService` (struct): passthrough unless message is targeted for our node's SE
- [x] `create_settlements_filter`
- [x] `core` (module)
- [x] `backends_common` (module)
- [x] `engines_api` (module)
- [x] `idempotency` (module)
- [x] `types` (module)
- [x] `NO_ENGINE_CONFIGURED_ERROR_TYPE` (const)
- [x] `CONVERSION_ERROR_TYPE` (const) 
- [x] `SE_ILP_ADDRESS` (const)
- [x] `Quantity` (struct)
- [x] `ApiResponse`(enum): Some APIs have default responses, e.g. Creating an account on the engine does not require us to get a response from the engine, so we use that to return a default value instead
- [x] `SettlementEngine` (trait): trait is documented, could use some extension and docs on its functions
- [x] `SettlementEngineDetails` (struct) 
- [x] `SettlementAccount` (trait)
- [x] `SettlementStore` (trait)
- [x] `LeftoversStore` (trait)
- [x] `ConvertDetails` (struct)
- [x] `Convert` (trait)
- [x] `scale_with_precision_loss` (fn)
- [x] `IdempotentStore` (trait)
- [x] `make_idempotent_call` (fn): Should this be made `pub(crate)`?
- [x] `get_hash_of` (fn) Should this be made `pub(crate)`? Note that this function is redefined in a few crates.
- [x] `CreateAccount`
- [x] `create_settlement_engine_filter`
- [x] `EngineRedisStore`
- [x] `EngineRedisStoreBuilder`

### PR DESCRIPTION
**Pull request by gakonst**
Pulled against #585 .

- [x] Closes #575 
- [x] Closes #574 
- [x] Closes #573 
- [x] Closes #572 
- [x] Closes #571 
- [x] Closes #570 
- [x] Closes #569  
- [x] Close[s] #576 WIP
- [x] service util docs
- [x] store docs
- [x] https://github.com/interledger-rs/interledger-rs/issues/449
Adds detailed docs for most things which were missing docs in the repo.

(CI is failing due to type length limit from #585, this is on my TODO list for the base)

### CODE REVIEW

File: crates/interledger-service-util/src/rate_limit_service.rs [hunk 1]
```diff
@@ -99,7 +114,23 @@ where
             .apply_rate_limits(request.from.clone(), request.prepare.amount())
             .await
         {
-            Ok(_) => next.handle_request(request).await,
+            Ok(_) => {
```
Comment (gakonst) [inline crates/interledger-service-util/src/rate_limit_service.rs]: This was a bug I introduced in the base PR

File: crates/interledger-store/src/redis/mod.rs [hunk 2]
```diff
@@ -272,8 +293,11 @@ impl RedisStoreBuilder {
 /// future versions of it will use PubSub to subscribe to updates.
 #[derive(Clone)]
 pub struct RedisStore {
+    //// The Store's ILP Address
     ilp_address: Arc<RwLock<Address>>,
+    /// A connection which reconnects if dropped by accident
     connection: RedisReconnect,
+    /// Websocket sender which publishes incoming payment updates
```
Comment (dora-gt) [inline crates/interledger-store/src/redis/mod.rs]:
```suggestion
    /// WebSocket sender which publishes incoming payment updates
```

File: crates/interledger-store/src/redis/reconnect.rs [hunk 3]
```diff
@@ -29,6 +29,7 @@ async fn get_shared_connection(redis_info: Arc<ConnectionInfo>) -> Result<Multip
 }
 
 impl RedisReconnect {
+    /// Connects to the provided redis connection
```
Comment (dora-gt) [inline crates/interledger-store/src/redis/reconnect.rs]:
```suggestion
    /// Connects to redis with the provided `ConnectionInfo`.
```

File: crates/interledger-stream/src/client.rs [hunk 4]
```diff
@@ -278,6 +310,11 @@ where
         Ok(())
     }
 
+    /// Parses the provided Fulfill packet.
+    /// 1. Logs the fulfill in the congestion controller
+    /// 1. Updates the last fulfill time of the send money future
+    /// 1. It tries to aprse a Stream Packet inside the fulfill packet's data field
```
Comment (dora-gt) [inline crates/interledger-stream/src/client.rs]:
```suggestion
    /// 1. It tries to parse a Stream Packet inside the fulfill packet's data field
```

File: crates/interledger-stream/src/client.rs [hunk 5]
```diff
@@ -278,6 +310,11 @@ where
...
+    ///    If successful, it icnrements the delivered amount by the Stream packet's prepare amount
```
Comment (dora-gt) [inline crates/interledger-stream/src/client.rs]:
```suggestion
    ///    If successful, it increments the delivered amount by the Stream packet's prepare amount
```

File: crates/interledger-stream/src/packet.rs [hunk 6]
```diff
@@ -105,23 +112,39 @@ impl<'a> StreamPacketBuilder<'a> {
...
+    /// 1. If the decryption fails
+    /// 1. If the decrypted bytes cannot be parsed to a unencrypted [Stream Packet](./struct.StreamPacket.html)
     pub fn from_encrypted(shared_secret: &[u8], ciphertext: BytesMut) -> Result<Self, ParseError> {
         // TODO handle decryption failure
         let decrypted = decrypt(shared_secret, ciphertext)
             .map_err(|_err| ParseError::InvalidPacket(String::from("Unable to decrypt packet")))?;
         StreamPacket::from_bytes_unencrypted(decrypted)
     }
 
+    /// Constructs a [Stream Packet](./struct.StreamPacket.html) from an buffer
```
Comment (dora-gt) [inline crates/interledger-stream/src/packet.rs]:
```suggestion
    /// Constructs a [Stream Packet](./struct.StreamPacket.html) from a buffer
```

File: crates/interledger-stream/src/packet.rs [hunk 7]
```diff
@@ -105,23 +112,39 @@ impl<'a> StreamPacketBuilder<'a> {
...
     frames_offset: usize,
 }
 
 impl StreamPacket {
+    /// Constructs a [Stream Packet](./struct.StreamPacket.html) from an encrypted buffer
+    /// and a shared secret
+    ///
+    /// # Errors
+    /// 1. If the decryption fails
+    /// 1. If the decrypted bytes cannot be parsed to a unencrypted [Stream Packet](./struct.StreamPacket.html)
```
Comment (dora-gt) [inline crates/interledger-stream/src/packet.rs]:
```suggestion
    /// 1. If the decrypted bytes cannot be parsed to an unencrypted [Stream Packet](./struct.StreamPacket.html)
```

File: crates/interledger-stream/src/packet.rs [hunk 8]
```diff
@@ -105,23 +112,39 @@ impl<'a> StreamPacketBuilder<'a> {
...
         // TODO handle decryption failure
         let decrypted = decrypt(shared_secret, ciphertext)
             .map_err(|_err| ParseError::InvalidPacket(String::from("Unable to decrypt packet")))?;
         StreamPacket::from_bytes_unencrypted(decrypted)
     }
 
+    /// Constructs a [Stream Packet](./struct.StreamPacket.html) from an buffer
+    ///
+    /// # Errors
+    /// 1. If the decrypted bytes cannot be parsed to a unencrypted [Stream Packet](./struct.StreamPacket.html)
```
Comment (dora-gt) [inline crates/interledger-stream/src/packet.rs]:
```suggestion
    /// 1. If the version of Stream Protocol doesn't match
    /// 1. If the provided bytes cannot be parsed to an unencrypted [Stream Packet](./struct.StreamPacket.html)
```

File: crates/interledger-stream/src/packet.rs [hunk 9]
```diff
@@ -105,23 +112,39 @@ impl<'a> StreamPacketBuilder<'a> {
...
+    /// The offset after which frames can be found inside the `buffer_unencrypted` field
     frames_offset: usize,
 }
 
 impl StreamPacket {
+    /// Constructs a [Stream Packet](./struct.StreamPacket.html) from an encrypted buffer
+    /// and a shared secret
+    ///
+    /// # Errors
+    /// 1. If the decryption fails
```
Comment (dora-gt) [inline crates/interledger-stream/src/packet.rs]:
```suggestion
    /// 1. If the version of Stream Protocol doesn't match
    /// 1. If the decryption fails
```

File: crates/interledger-stream/src/packet.rs [hunk 10]
```diff
@@ -460,9 +503,13 @@ impl<'a> fmt::Debug for ConnectionNewAddressFrame {
     }
 }
 
+/// The assets being transported in this Stream payment
+/// Asset details exposed by this frame MUST NOT change during the lifetime of a Connection.
```
Comment (dora-gt) [inline crates/interledger-stream/src/packet.rs]: Interesting. Can we ensure that now?

File: crates/interledger-http/src/client.rs [hunk 11]
```diff
@@ -12,10 +12,20 @@ use reqwest::{
 use secrecy::{ExposeSecret, SecretString};
 use std::{convert::TryFrom, marker::PhantomData, sync::Arc, time::Duration};
 
+/// The HttpClientService implements [OutgoingService](../../interledger_service/trait.OutgoingService)
+/// for sending ILP Prepare packets over to the HTTP URL associated with the provided account
+/// If no [ILP-over-HTTP](https://interledger.org/rfcs/0035-ilp-over-http) URL is specified for
+/// the account in the request, then it is forwarded to the next service.
```
[the remark this answers is not in the document]
Comment (gakonst) [reply crates/interledger-http/src/client.rs]: Don't think we should write that, as this is just our configuration. Maybe somebody else will create a node which does not use the BTP service

File: crates/interledger-settlement/src/core/types.rs [hunk 12]
```diff
@@ -45,62 +54,100 @@ impl Quantity {
...
         idempotency_key: Option<String>,
     ) -> Result<(), ()>;
 
+    /// Increases the account's balance by the provided amount.
+    /// Only call this if a settlement request has failed
     async fn refund_settlement(&self, account_id: Uuid, settle_amount: u64) -> Result<(), ()>;
 }
 
+/// Trait used by the connector and engine to track amounts which should have been
+/// settled but were not due to precision loss
```
[the remark this answers is not in the document]
Comment (gakonst) [reply crates/interledger-settlement/src/core/types.rs]: In the settlement spec, https://github.com/interledger/rfcs/blob/master/0038-settlement-engines/0038-settlement-engines.md

File: crates/interledger-stream/src/packet.rs (hunk 10 above)
Comment (gakonst) [reply crates/interledger-stream/src/packet.rs]: We do not check that this does not happen, but we never change it on our side

**Comment by gakonst** [comment]
Closing in favor of new stacked PRs.

File: crates/interledger-service-util/src/rate_limit_service.rs (hunk 1 above)
Comment (gakonst) [reply crates/interledger-service-util/src/rate_limit_service.rs]: https://github.com/interledger-rs/interledger-rs/issues/610

**Pull request closed without merging**